Add Map/Reduce Message

hazen opened the pull request:
Queue up behind other changes, but basically is simple M/R implementation.

Commit:
Initial pass at adding Map/Reduce message

Commit:
Merge branch 'develop' into HEAD

Commit:
Merge branch 'develop' into HEAD

broach commented:
When performing a map-reduce, if there are not sufficient vnodes, riak will reply: 

`{"error":"insufficient_vnodes_available"}` 

This causes the client to segfault. 

**Edit:** In fact, any error will. Invalid JSON, JS, etc.

Commit:
Don't free map reduce response on error

The client example frees the map reduce example even when
err != ERIAK_OK causing a segfault.

broach commented:
After fixing problem in client example, this works as expected. the map-reduce result is returned correctly or an error is reported. I replicated the example at http://docs.basho.com/riak/latest/dev/using/mapreduce/ using the `riak_c_example` 

:+1:
